src/sparse/impl/KokkosSparse_gauss_seidel_impl.hpp

```diff
@@ -887,6 +912,340 @@ namespace KokkosSparse{
+        }
+        
+        ColorView colors;
+        color_t numClusterColors;
+        nnz_lno_t numClusters;
+      };
+
+      struct ClusterToVertexColoring
+      {
+        typedef typename HandleType::GraphColoringHandleType GCHandle;
+        typedef typename GCHandle::color_view_t ColorView;
+
+        ClusterToVertexColoring(ColorView& clusterColors_, ColorView& vertexColors_, nnz_lno_t numRows_, nnz_lno_t numClusters_, nnz_lno_t clusterSize_)
+          : clusterColors(clusterColors_), vertexColors(vertexColors_), numRows(numRows_), numClusters(numClusters_), clusterSize(clusterSize_)
+        {}
+
+        KOKKOS_INLINE_FUNCTION void operator()(const size_type i) const
+        {
+          size_type cluster = i / clusterSize;
+          size_type clusterOffset = i - cluster * clusterSize;
+          vertexColors(i) = ((clusterColors(cluster) - 1) * clusterSize) + clusterOffset + 1;
+        }
+
+        ColorView clusterColors;
+        ColorView vertexColors;
+        nnz_lno_t numRows;
+        nnz_lno_t numClusters;
+        nnz_lno_t clusterSize;
+      };
+
+      template<typename rowmap_t, typename colinds_t>
+      typename HandleType::GraphColoringHandleType::color_view_t initialize_symbolic_cluster(rowmap_t xadj, colinds_t adj, color_t& numColors)
+      {
+#if KOKKOSSPARSE_IMPL_PRINTDEBUG
+        Kokkos::Impl::Timer timer;
+#endif
+        typename HandleType::GaussSeidelHandleType *gsHandler = this->handle->get_gs_handle();
+        typedef typename HandleType::GraphColoringHandleType::color_view_t color_view_t;
+        //compute the RCM ordering of the graph
+        RCM<HandleType, rowmap_t, colinds_t> rcm(num_rows, xadj, adj);
+        //rcmOrder maps (bijectively) from original rows to RCM-ordered rows.
+        //rcmOrder[i] represents where in the _order_ row i is.
+        //rcmPerm is the inverse mapping: a list of all rows _permuted_ into the RCM ordering.
+        nnz_lno_persistent_work_view_t rcmOrder = rcm.rcm();
+        nnz_lno_persistent_work_view_t rcmPerm("RCM permutation array", num_rows);
+        Kokkos::parallel_for(my_exec_space(0, num_rows), OrderToPermFunctor(rcmOrder, rcmPerm));
+#if KOKKOSSPARSE_IMPL_PRINTDEBUG
+        std::cout << "\nRCM:" << timer.seconds() << '\n';
+        timer.reset();
+#endif
+        auto clusterSize = gsHandler->get_cluster_size();
+        nnz_lno_t numClusters = (num_rows + clusterSize - 1) / clusterSize;
+        //build the cluster graph using the (implicitly permuted) RCM order of the matrix (xadj, adj)
+        //first, count the entries per row
+        const size_type bitsPerST = 8 * sizeof(size_type);
+        const size_type wordsPerRow = (numClusters + bitsPerST - 1) / bitsPerST;
+        size_type nthreads = MyExecSpace::concurrency();
+        if(nthreads > 64)
+          nthreads = 64;
+#ifdef KOKKOS_ENABLE_CUDA
+        if(std::is_same<MyExecSpace, Kokkos::Cuda>::value)
+        {
+          nthreads = 256;
```

Comment:
Is this correct ? Do you want this many Cudathreads ?

Reply:
@srajama1 This code to build the cluster graph using fixed-size dense accumulators has all been replaced with a much better version.